server/src/test/java/org/elasticsearch/index/IndexServiceTests.java

```diff
@@ -382,6 +384,36 @@ public void testAsyncTranslogTrimActuallyWorks() throws Exception {
         assertBusy(() -> assertThat(IndexShardTestCase.getTranslog(shard).totalOperations(), equalTo(0)));
     }
 
+    public void testAsyncTranslogTrimTaskOnClosedIndex() throws Exception {
+        final String indexName = "test";
+        IndexService indexService = createIndex(indexName, Settings.builder()
+            .put(TRANSLOG_RETENTION_CHECK_INTERVAL_SETTING.getKey(), "100ms")
+            .build());
+
+        final Translog translog = IndexShardTestCase.getTranslog(indexService.getShard(0));
+        final int numDocs = scaledRandomIntBetween(10, 100);
+        for (int i = 0; i < numDocs; i++) {
+            client().prepareIndex().setIndex(indexName).setId(String.valueOf(i)).setSource("{\"foo\": \"bar\"}", XContentType.JSON).get();
+            if (randomBoolean()) {
+                client().admin().indices().prepareFlush(indexName).get();
+            }
+        }
+        assertAcked(client().admin().indices().prepareClose("test"));
+
+        indexService = getInstanceFromNode(IndicesService.class).indexServiceSafe(indexService.index());
+        assertTrue(indexService.getTrimTranslogTask().mustReschedule());
+
+        final long lastCommitedTranslogGeneration;
+        try (Engine.IndexCommitRef indexCommitRef = getEngine(indexService.getShard(0)).acquireLastIndexCommit(false)) {
+            Map<String, String> lastCommittedUserData = indexCommitRef.getIndexCommit().getUserData();
+            lastCommitedTranslogGeneration = Long.parseLong(lastCommittedUserData.get(Translog.TRANSLOG_GENERATION_KEY));
+        }
+        assertBusy(() -> {
+            long minTranslogGen = Translog.readMinTranslogGeneration(translog.getConfig().getTranslogPath(), translog.getTranslogUUID());
+            assertThat(minTranslogGen, equalTo(lastCommitedTranslogGeneration));
+        });
```

Comment:
Can we reopen the index and verify that there's no operation in TranslogStats?

Reply:
I pushed fe62689dd6a3c11c6a647cbaea9ed5bce221b706